staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -989,31 +1074,53 @@ func (c *cacheWatcher) add(event *watchCacheEvent, timer *time.Timer, budget *ti
 	budget.returnUnused(timeout - time.Since(startTime))
 }
 
-// NOTE: sendWatchCacheEvent is assumed to not modify <event> !!!
-func (c *cacheWatcher) sendWatchCacheEvent(event *watchCacheEvent) {
-	curObjPasses := event.Type != watch.Deleted && c.filter(event.Key, event.ObjLabels, event.ObjFields)
-	oldObjPasses := false
-	if event.PrevObject != nil {
-		oldObjPasses = c.filter(event.Key, event.PrevObjLabels, event.PrevObjFields)
-	}
-	if !curObjPasses && !oldObjPasses {
-		// Watcher is not interested in that object.
-		return
+func (c *cacheWatcher) addBookmark(event *watchCacheEvent) {
+	// send the bookmark event when the client support it
+	if c.bookmarkSendingPeriod() > 0 {
+		// If we can't send it, don't block on it.
+		select {
+		case c.input <- event:
+		default:
+		}
 	}
+}
+
+// It seems that sending bookmark to every watcher every X seconds is too expensive,
+// which make good-enough workaround is to send bookmark right before watch timeout.
+// FIXME let's leave it for now and we will add that (if needed) later
+func (c *cacheWatcher) bookmarkSendingPeriod() time.Duration {
```

Comment:
I think that the interface for this function isn't exactly what we want for extensibility - I don't think that bookmarks should in general be send in equal periods.
I think that instead what we want is:
```
func (c *cacheWatcher) nextBookmarkTime(now time.Time) time.Time {
  // For now we return timeout - 2s (and maybe +infinity is now already passed this time)
  // but it gives us extensibility for the future.
}
```

If at some point we will decide to send bookmarks every e.g. 1m, we still want to send it right before timeout anyway, so we need some flexibility.

Reply:
Agree with the proposal.